### PR TITLE
Use the new format to create MISP events

### PR DESCRIPTION
**Pull request by Rafiot**
You should see massive speed improvements as it does one single query per new MISP event.

Please give it a try, I have no feed to test with, so it very probably contains a few bugs.

### CODE REVIEW

**Comment by github-germ** [comment]
@Rafiot -- Excellent!! Thanks...

I'm working on prepping our first MISP deployment for tomorrow. Once I get some cycles (soon!) I will test and report back. I will also apply these changes to the enhanced version I now have as I understand the changes.

(Note: As I mentioned on gitter, we've initiated our corporate process granting "permissions" so we can share back to the project. Not sure how long that will take, but wanted you to know our intent.)

**Comment by github-germ** [comment]
I did not try the new `PySight.py` yet (because it will likely not work since the original didn't, and I had test case code for creating events with hashes already that reproduced the 503 issue previously). However, I revised my test case code I wrote which does NOT use threads but rather separate processes that each do the same:
1. populate an in-memory `MISPEvent`
2. create a `MISPObject`and calls `add_attribute` 4 times for `object_relation filename, md5, sha1, sha256` followed by `add_object` to that `MISPEvent`.
3. send that `MISPEvent` up to the server via `add_event`

When I run that test case with only one process, it works fine: the event looks fine in the WebUI. However, when I run with 10 processes, things go bad (not sure what the breaking point is for number of processes, but 10 is reliably failing, and will vary depending on the config I think):
* Exception to my client test code
```
CRITICAL [api.py:215 - _check_response() ] Something bad happened
on the server-side: {'name': 'An Internal Error Has Occurred.',
'message': 'An Internal Error Has Occurred.', 'url': '/events'}
```
* Apache `misp.local_access.log` HTTP 500
```
::1 - - [20/Jun/2018:00:07:34 +0000] "POST /events HTTP/1.1" 500 103
 "-" "PyMISP 2.4.92.1 - Python 3.6.4"
```
* And here's one trace from `/var/www/MISP/app/tmp/logs/error.log`:
```
2018-06-20 00:07:43 Error: [PDOException] SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction
Request URL: /events
Stack Trace:
#0 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(472): PDOStatement->execute(Array)
#1 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(438): DboSource->_execute('INSERT INTO `mi...', Array)
#2 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1092): DboSource->execute('INSERT INTO `mi...')
#3 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create(Object(AppModel), Array, Array)
#4 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Model.php(1754): Model->_doSave(Array, Array)
#5 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Model.php(2352): Model->save(Array, Array)
#6 /var/www/MISP_2.4.92/app/Model/Attribute.php(1797): Model->saveMany(Array, Array)
#7 /var/www/MISP_2.4.92/app/Model/Attribute.php(582): Attribute->__afterSaveCorrelation(Array)
#8 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Event/CakeEventManager.php(241): Attribute->afterSave(true, Array)
#9 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Model.php(1970): CakeEventManager->dispatch(Object(CakeEvent))
#10 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave(Array, Array)
#11 /var/www/MISP_2.4.92/app/Model/Attribute.php(3135): Model->save(Array, Array)
#12 /var/www/MISP_2.4.92/app/Model/MispObject.php(503): Attribute->captureAttribute(Array, '12295', Array, '6930', Object(Log))
#13 /var/www/MISP_2.4.92/app/Model/Event.php(2693): MispObject->captureObject(Array, '12295', Array, Object(Log))
#14 /var/www/MISP_2.4.92/app/Controller/EventsController.php(1409): Event->_add(Array, true, Array, '', NULL, false, NULL, 0, Array)
#15 [internal function]: EventsController->add()
#16 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Controller/Controller.php(491): ReflectionMethod->invokeArgs(Object(EventsController), Array)
#17 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction(Object(CakeRequest))
#18 /var/www/MISP_2.4.92/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke(Object(EventsController), Object(CakeRequest))
#19 /var/www/MISP_2.4.92/app/webroot/index.php(92): Dispatcher->dispatch(Object(CakeRequest), Object(CakeResponse))
#20 {main}
```

**Comment by github-germ** [comment]
This test case code has grown somewhat unwieldy as you can specify various permutations of execution, but it gets the job done. Depending on your config, you might need to run more concurrent processes, e.g. 2 didn't always fail for me, but 10 does for my setup.

Two files:

1. `testInMemory.sh` -- use this to drive the test. The key is play with line 25 which is not set for each process to to create 2 events, with 100 hash groups (file, md5, sh1, sha256) for each event, creating them in-memory objects, within one process use only one `PyMISP` connection, vary the hashes. Run this script with the argument of how many processes you want, e.g. `testInMemory.sh 10` -- that will create 10 log files where you will see the `something bad` errors.
2. `mispClientInMemory.py` -- the script for each process as called from `testInMemory.sh`.

[PyMISP_AddHashesTestCase.zip](https://github.com/deralexxx/PySight2MISP/files/2117386/PyMISP_AddHashesTestCase.zip)

Before I hack my dramatically revised version of `PySight.py` to this in-memory model, it would be a big help if you could review the essence of what this test code does, and let me know if I did something wrong, or if there still is an issue.

Many thanks!!

**Comment by github-germ** [comment]
Btw, I also previously saw 503s with a single threaded loop merely doing consecutive `delete_event`. Is this somehow related? See https://github.com/MISP/MISP/issues/3305#issuecomment-395443896

And when I try to delete my test case created events with the WebUI, the WebUI takes a few minutes and then comes back to the browser with a 503. The one I just tried was 60 events, many with 200-400 attributes. Didn't see anything in the MISP `error.log`. When I refreshed the `List Events` and reused the filter to find my events, there were 12 remaining out of the 60. I then select those, and they deleted OK.

**Comment by Rafiot** [comment]
I do not have a feed to test, so it will be difficult for me to give you much more support.

Just a practical question still: is multithreading/multiprocessing required with the new approach? How many events are you receiving and trying to create per sec/min? 
The new approach should be dramatically faster already.

And note than the log entries starting with `Something bad happened on the server-side:` mean than the server didn't return any useful information and PyMISP doesn't know what to do with it.

**Comment by github-germ** [comment]
Hi, this test case code does not use any feed data. It’s self contained so you can use it.

**Comment by github-germ** [comment]
Can you explain why the simple loop or the WebUI deleting events gets 503s?

**Comment by github-germ** [comment]
`Just a practical question still: is multithreading/multiprocessing required with the new approach? How many events are you receiving and trying to create per sec/min?
The new approach should be dramatically faster already.`

I can refector my code but was looking to make sure the issue would be resolved before doing so. We can get 5k IOCs. I can see about it performance when I change the code. However, over time as we continue to grow the number of feeds, I’m still concerned.

**Comment by Rafiot** [comment]
I have no doubts you have issues with MISP and multiprocessing, and it is tricky to solve. My approach right now is to find a way that works fast enough, and reliably, to ingest your feeds. 

5K IOCs, in a day? If yes, you definitely don't need multiprocessing: it will be done within a few minutes with the current script. If it is 5k every minute, I can give you different recommendations (such as disabling correlation). 

The simple loop with delete is a problem we need to have a look with @iglocska that is not directly related to this specific multiprocessing issue, we will look at it on the other issue you opened.

**Comment by github-germ** [comment]
1. Re' `simple loop delete problem`: I created a separate issue: https://github.com/MISP/MISP/issues/3361 -- HTTP 503 with many Event Deletions
2. I will start refactoring my script and report back when that's tested. Might take a bit due to other items on my plate.
3. Re' MISP and mulit-processing challenges in general: if it's not my one multi-threaded ingestion script "stressing" the server-side, what do organizations that may ingest through a variety of methods from a multitude of sources do if they can not time-spread the ingestion without overlapping concurency causing the same overload conditions?

**Comment by iglocska** [comment]
I have a feeling it IS directly related. Even if your python script is single threaded, if it hammers away at an FPM enabled server without waiting for responses it will get 503s once FPM runs out of threads. 

tl;dr: your client is not multithreaded, the server is. You trigger heavy multithreading with your single thread client.

**Comment by github-germ** [comment]
**Questions** for @iglocska where in https://github.com/deralexxx/PySight2MISP/pull/6#issuecomment-399013982 he wrote: 

> ... it hammers away at an FPM [1] **enabled server without waiting for responses** it will get 503s once [2] **FPM runs out of threads**.

(Forgive me as I don't know FPM)

1.  Is there a way to force FPM (or something else in the pipeline) to wait for responses?
2. My threaded code has a configurable high water active thread limit. How can we config FPM the support at least that number of threads?

**Comment by iglocska** [comment]
Basically, if I understand what is going wrong it's this:

1. No idea.
2. If I understand what is going wrong: your threaded code's limits don't
matter. Each time it queries FPM you'll initialise a process on FPM that
will start execution. Your script will keep query-ing away without waiting
for responses.

**Comment by github-germ** [comment]
replying to `Your script will keep query-ing away without waiting for responses.`

Are you saying that `PyMISP` does not wait for completion of a request from the server-side before returning back to the client caller code and there is no way for the client to block or poll for completion?

**Comment by github-germ** [comment]
@Rafiot ... in https://github.com/deralexxx/PySight2MISP/pull/6/commits/0ec40b5e15c84ec1705fd7dc6018395fb338d677 should all the `new_misp_event.tag(tag=...)` actually be `new_misp_event.add_attribute_tag(tag=...)` or `new_misp_event.add_attribute(type=..., value=...)` or perhaps something else?

I ask because of this runtime error: `'MISPEvent' object has no attribute 'tag'`

**Comment by Rafiot** [comment]
Good point, the right call is `add_tag`

File: PySight.py [hunk 1]
```diff
@@ -477,86 +464,80 @@ def is_map_alert_to_event(p_misp_instance, a_event, a_isight_alert, a_auto_comme
...
         if a_isight_alert.emailLanguage:
-            p_misp_instance.add_internal_other(a_event, a_isight_alert.emailLanguage, False,
-                                               "E-mail language " + a_auto_comment)
-    except TypeError as e:
+            attr = new_misp_event.add_attribute(type='other', value=a_isight_alert.emailLanguage, category='Internal reference', comment='E-mail language {}'.format(a_auto_comment))
+            if has_email:
+                new_email_object.add_reference(attr.uuid, 'related-to', 'E-mail language')
+        if has_email:
+            new_misp_event.add_object(new_email_object)
+        p_misp_instance.add_event(new_misp_event)
```
Comment (github-germ) [inline PySight.py:526]: in my version of this revised code when finally sending the event up to the server here, i get this in the `error.log`. where should `info` be set?
```
Error: [PDOException] SQLSTATE[HY000]: General error: 1364 Field 'info' doesn't have a default value Request URL: /events
```

**Comment by github-germ** [comment]
@Rafiot -- hope you can help with this error: https://github.com/deralexxx/PySight2MISP/pull/6/files#diff-064503d379390dc031f013232f9d9354R528

File: PySight.py [hunk 2]
```diff
@@ -617,7 +598,7 @@ def misp_check_for_previous_events(misp_instance, isight_alert):
             try:
                 event = misp_instance.new_event(0, 2, 0, isight_alert.title + " pySightSight " + isight_alert.reportId,
```
Comment (github-germ) [inline PySight.py]: A bit confused here. We are creating a `new_event` here and adding to the server vs. doing the in-memory construction of the `MISPEvent` in `is_map_alert_to_event` (https://github.com/deralexxx/PySight2MISP/pull/6/files#diff-064503d379390dc031f013232f9d9354R349)

And there 2 more calls to `new_event` just below here.

**Comment by Rafiot** [comment]
Okay, seems I missed a few parts of the script. I'll have a look today or over the weekend.

File: PySight.py (hunk 1 above)
Comment (Rafiot) [reply PySight.py:526]: That should be fixed in the commit I'll do in a few min.

File: PySight.py (hunk 2 above)
Comment (Rafiot) [reply PySight.py]: new_event were mistakes

**Review by github-germ — APPROVED** [review]
Good. I did something like this last night in my code. Thx.

**Comment by github-germ** [comment]
@Rafiot ... OK, firstly, thanks again for all your help on this issue.

I've refactored my version of the iSIGHT ingestion to do in-memory MISPEvent construction. It is working without errors most of the time. However, when running muti-threaded `update_event` or `add_event` can still experience deadlock:
```
Error: [PDOException] SQLSTATE[40001]: Serialization failure: 1213 Deadlock \
found when trying to get lock; try restarting transaction
```

If I run it in single-threaded (no threads) mode, no errors, but it does run slower. This may be workable for us -- need to run benchmarks on a test host with the same iron as our production (I develop on a skinnier VM).

The good news: hashes now work (didn't work even single-threaded before), the server is being hit less, and single-threaded now is faster than single-threaded before.

QUESTION:
1. Can you elaborate on what exactly the `asynch` argument to the `PyMISP` class does? I see it defaults to `False`. Is it possible that by setting it `True` the db deadlock might resolve? (My intuition says it would work better `False`, which is what we have, but wanted to ask).

**Comment by Rafiot** [comment]
`async` is very untested and is supposed to allow you to run async queries against the remote instance: https://docs.python.org/3/library/asyncio-task.html

I strongly recommend you to leave it to `False`.

**Comment by github-germ** [comment]
OK, so do you agree that running ingestion clients in general multi-threaded is not advisable with the current MISP?

And if so, are there any future plans to making the server-side more robust to support concurrent client accessing common data elements (either multiple processes on the same or different client hosts or a multi-threaded client on one host hitting the server)?

Thx again for all your help!

**Comment by Rafiot** [comment]
Short answer: yes, we will improve the ingestion speed of MISP, but I cannot tell you when and how, nor if we will ever recommend the use of multithreading.

Long answer: Our approach is to "not fix a bug if it ain't one". Is the current script fast enough to import your feed? If yes, this bug is fixed on our side. If not, we still have an issue and we need to fix it.

The reason I say that is that you will always be able to hammer the system enough so it fails randomly it can be with multithreading, pushing huge events, running MISP on a rPi, ... The script I propose here uses the recommended way to interact with MISP, making it (hopefully) working reliably, and that's our goal.

**Comment by github-germ** [comment]
@Rafiot -- OK, my refactored (greatly enhanced) version of `PySight.py` using your version as a model works reliably single-threaded (i.e. no threads) and appears to complete within reasonable time constraints for our needs.

**THANK YOU VERY MUCH!**

I will add that I've not tested the version you created as it does not work as is. When I first brought in that version in May, I had to make some changes just to get it to work (even with the prior method of posting events to MISP via `PyMISP`). Since then I've greatly enhanced that baseline version, and now have refactored it to your recommended in-memory `MISPEvent` approach to adding/updating events via `PyMISP`.

That said, if I get permission from my team to share my code here, what's the best way to do that? I still use the same file names in my version. (I've not committed/pushed here to date)

**Comment by Rafiot** [comment]
Great news! 

If you can share the code, the best way is to add it on github under your name (just create a new repo). Or if it makes sense, do a pull request in this repository.

**Comment by jaegeral** [comment]
Happy to merge a pull requestAm 05.07.2018 15:46 schrieb Raphaël Vinot <notifications@github.com>:Great news!
If you can share the code, the best way is to add it on github under your name (just create a new repo). Or if it makes sense, do a pull request in this repository.
